### DISCUSSION TITLE
[iOS] Can you receive/manage data-only messages when the app is closed?

### DISCUSSION THREAD

**Discussion by piljac1** [invertase/react-native-firebase#5485] [Q&A] [answered]
Hi! I'm opening this discussion because I've spent hours trying to find a definitive answer, but all I found is that some people are saying that it is possible and some are saying that it isn't. As for myself, I haven't been able to make `setBackgroundMessageHandler` work when the app is closed (but it works when the app is in background). When I send a data-only message through FCM/APNs, I can see that my app starts in background (headless), but nothing happens. My `setBackgroundMessageHandler` handler is not called. I also tried implementing `getInitialNotification`, but it always returned undefined. From my understanding, you need an actual notification payload to leverage it?

Anyway, thanks in advance!

P.S. If it turns out it is feasible, I'll gladly post some code to try and understand what's wrong.

**Comment by mikehardy** [comment, marked as answer]
a) "closed" == just in the background on iOS / not visible, or 
b) "closed" == you flicked the app away from the app chooser / force quit it?

a) you *can* receive data-only messages, but they will not be reliably delivered. Do not design a solution where delivery is required here
b) nope, iOS listens to the user: the user asked for the app to stop running so the operating system will not run the app in response to FCM

**Reply by piljac1** [reply]
In my case I was talking about b). We started by using notifications, but then I thought about a possible edge. We have a settings screen in our app from which you can enable/disable push notifications and when you enable them, it subscribes the device to a topic and if you disable them, it unsubscribes the device. If you go offline while having your push notifications enabled and then disable them, the unsubscribe action will be cached until the device is online again. However, if push notifications are sent within this time frame, they will be received when the device is back online and then, the unsubscribe event will be processed and the user won't receive any other notifications after that. I thought that it might be confusing for the user to disable notifications and then receive notifications when you go back online. So that's why we were headed towards data-only messages (so we can run some business logic to avoid sending unnecessary push notifications in that context). In the end, it's a really far fetched edge case and if fixing that edge case means compromising notification deliveries, I guess we'll live with it.

**Reply by mikehardy** [reply]
iOS notification extensions are the answer I think? a la https://notifee.app/react-native/docs/ios/remote-notification-support (of course, that's a licensed product but it shows the style / possibilities - notification block in FCM implies reliable delivery, notification extension implies a brief moment of app control prior to local display)
